File: ruby/bob/bob.rb
class Utterance

  def initialize(string)
    @string = string
  end

  def type
    # check shout before question because a shouted question counts as a
    # shout, not a question.  note that even if it ends in "!", it's not
    # a shout unless all letters are uppercase.  (ignore non-letters.)
    if all_chars_uppercase
      :shout
    elsif @string.end_with? "?"
      :question 
    elsif ! @string.strip.empty?
      :statement
    else
      :empty
    end
  end

  def all_chars_uppercase
    @string =~ /[[:upper:]]/ && @string !~ /[[:lower]]/
  end

end


class Bob

  def hey(utterance)
    type = Utterance.new(utterance).type
    type = :unknown unless RESPONSES.has_key? type
    RESPONSES[type]
  end

  private

  RESPONSES = {
    statement: "Whatever.",
    shout:     "Whoa, chill out!",
    question:  "Sure.",
    empty:     "Fine. Be that way!",
    unknown:   "Huh?"
  }

end

File: elixir/bob/bob.exs
defmodule Bob do

  @spec hey(String.t) :: String.t()
  def hey(input) do
    cond do
      # order is important here; a shouted question
      # is considered a question, not a shout.
      # (USED to be considered a shout though.  Harrumph.)
      silence?(input)  -> "Fine. Be that way!"
      question?(input) -> "Sure."
      shout?(input)    -> "Whoa, chill out!"
      true             -> "Whatever."
    end
  end


  defp silence?(input) do
    String.rstrip(input) == ""
  end

  defp shout?(input) do
    # need to ensure there is at least one letter; "3" is not a shout,
    # even though it matches its upcased version.
    # credit to someone else at the meetup for the alpha regex shortcut.
    String.upcase(input) == input && input =~ ~r/[[:alpha:]]/
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

end
